Backend Internal error: Exception during psi2ir when using with kotlin 1.8.20

After upgrading kotlin to 1.8.20 i have this error during compilation:

```
org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during psi2ir
File being compiled: (25,84) in /Users/jakubsyty/Projects/AndroidArchitectureTemplate/app/build/anvil/src-gen-debug/pl/jsyty/architecturetemplate/GeneratedApplicationComponent.kt
The root cause org.jetbrains.kotlin.psi2ir.generators.ErrorExpressionException was thrown at: org.jetbrains.kotlin.psi2ir.generators.ErrorExpressionGenerator.generateErrorCall(ErrorExpressionGenerator.kt:100)
null: KtCallExpression:
factory()
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:253)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException$default(CodegenUtil.kt:237)
	at org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator.generateMemberDeclaration(DeclarationGenerator.kt:75)
	at org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator.generateClassMemberDeclaration(DeclarationGenerator.kt:101)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateMembersDeclaredInClassBody(ClassGenerator.kt:491)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateClass(ClassGenerator.kt:101)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateClass$default(ClassGenerator.kt:63)
	at org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator.generateClassOrObjectDeclaration(DeclarationGenerator.kt:108)
	at org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator.generateMemberDeclaration(DeclarationGenerator.kt:53)
	at org.jetbrains.kotlin.psi2ir.generators.ModuleGenerator.generateSingleFile(ModuleGenerator.kt:73)
	at org.jetbrains.kotlin.psi2ir.generators.ModuleGenerator.generateModuleFragment(ModuleGenerator.kt:53)
	at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment(Psi2IrTranslator.kt:97)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.convertToIr(JvmIrCodegenFactory.kt:224)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.convertToIr(JvmIrCodegenFactory.kt:57)
	at org.jetbrains.kotlin.codegen.KotlinCodegenFacade.compileCorrectFiles(KotlinCodegenFacade.java:41)
	at org.jetbrains.kotlin.kapt3.AbstractKapt3Extension.contextForStubGeneration(Kapt3Extension.kt:280)
	at org.jetbrains.kotlin.kapt3.AbstractKapt3Extension.analysisCompleted(Kapt3Extension.kt:174)
	at org.jetbrains.kotlin.kapt3.ClasspathBasedKapt3Extension.analysisCompleted(Kapt3Extension.kt:104)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM$analyzeFilesWithJavaIntegration$2.invoke(TopDownAnalyzerFacadeForJVM.kt:115)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration(TopDownAnalyzerFacadeForJVM.kt:125)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration$default(TopDownAnalyzerFacadeForJVM.kt:99)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:257)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:42)
	at org.jetbrains.kotlin.cli.common.messages.AnalyzerWithCompilerReport.analyzeAndReport(AnalyzerWithCompilerReport.kt:115)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.analyze(KotlinToJVMBytecodeCompiler.kt:248)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.repeatAnalysisIfNeeded(KotlinToJVMBytecodeCompiler.kt:182)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:88)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:47)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:168)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:53)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:100)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:46)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:101)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1486)
	at jdk.internal.reflect.GeneratedMethodAccessor106.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.jetbrains.kotlin.psi2ir.generators.ErrorExpressionException: null: KtCallExpression:
factory()
	at org.jetbrains.kotlin.psi2ir.generators.ErrorExpressionGenerator.generateErrorCall(ErrorExpressionGenerator.kt:100)
	at org.jetbrains.kotlin.psi2ir.generators.StatementGenerator.visitCallExpression(StatementGenerator.kt:390)
	at org.jetbrains.kotlin.psi2ir.generators.StatementGenerator.visitCallExpression(StatementGenerator.kt:54)
	at org.jetbrains.kotlin.psi.KtCallExpression.accept(KtCallExpression.java:35)
	at org.jetbrains.kotlin.psi2ir.generators.StatementGenerator.visitDotQualifiedExpression(StatementGenerator.kt:427)
	at org.jetbrains.kotlin.psi2ir.generators.StatementGenerator.visitDotQualifiedExpression(StatementGenerator.kt:54)
	at org.jetbrains.kotlin.psi.KtDotQualifiedExpression.accept(KtDotQualifiedExpression.kt:32)
	at org.jetbrains.kotlin.psi2ir.generators.StatementGenerator.generateStatement(StatementGenerator.kt:70)
	at org.jetbrains.kotlin.psi2ir.generators.StatementGenerator.generateExpression(StatementGenerator.kt:83)
	at org.jetbrains.kotlin.psi2ir.generators.BodyGenerator.generateExpressionBody(BodyGenerator.kt:93)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateDelegatedImplementationMembers(ClassGenerator.kt:220)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateMembersDeclaredInSupertypeList(ClassGenerator.kt:192)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateClass(ClassGenerator.kt:99)
	at org.jetbrains.kotlin.psi2ir.generators.ClassGenerator.generateClass$default(ClassGenerator.kt:63)
	at org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator.generateClassOrObjectDeclaration(DeclarationGenerator.kt:108)
	at org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator.generateMemberDeclaration(DeclarationGenerator.kt:53)
	... 47 more
```

Generated code causing that:

```java
// Automatically generated file. DO NOT MODIFY
@file:Suppress(
  "DEPRECATION",
  "OPT_IN_USAGE",
  "OPT_IN_USAGE_ERROR",
)

package pl.jsyty.architecturetemplate

import com.deliveryhero.whetstone.SingleIn
import com.deliveryhero.whetstone.app.ApplicationComponent
import com.deliveryhero.whetstone.app.ApplicationScope
import com.squareup.anvil.annotations.MergeComponent
import dagger.Component
import javax.inject.Singleton
import kotlin.Suppress

@MergeComponent(ApplicationScope::class)
@SingleIn(ApplicationScope::class)
@Singleton
public interface GeneratedApplicationComponent : ApplicationComponent {
  @Component.Factory
  public interface Factory : ApplicationComponent.Factory

  public companion object Default : Factory by DaggerGeneratedApplicationComponent.factory()
}
```

Offending line is: `public companion object Default : Factory by DaggerGeneratedApplicationComponent.factory()`, specially the last `factory()` call.

Repro is here: https://github.com/jakoss/AndroidArchitectureTemplate/pull/85

Duplicate of this issue is posted on Anvil repo, but they are not sure if it's related to their work or Whetstone: https://github.com/square/anvil/issues/702

Maybe kotlin and Anvil version upgrade could help fix the issue?

--- a/whetstone-compiler/src/main/java/com/deliveryhero/whetstone/compiler/FqNames.kt
+++ b/whetstone-compiler/src/main/java/com/deliveryhero/whetstone/compiler/FqNames.kt
@@ -10,4 +10,5 @@ internal object FqNames {
     @JvmField val AUTO_INJECTOR = FqName("com.deliveryhero.whetstone.meta.AutoInjectorBinding")
     @JvmField val AUTO_INSTANCE = FqName("com.deliveryhero.whetstone.meta.AutoInstanceBinding")
     @JvmField val SINGLE_IN = FqName("com.deliveryhero.whetstone.SingleIn")
+    @JvmField val APPLICATION = FqName("android.app.Application")
 }

--- a/whetstone-compiler/src/main/java/com/deliveryhero/whetstone/compiler/handlers/AppComponentHandler.kt
+++ b/whetstone-compiler/src/main/java/com/deliveryhero/whetstone/compiler/handlers/AppComponentHandler.kt
@@ -2,6 +2,7 @@ package com.deliveryhero.whetstone.compiler.handlers
 
 import com.deliveryhero.whetstone.compiler.CodegenHandler
 import com.deliveryhero.whetstone.compiler.FqNames
+import com.deliveryhero.whetstone.compiler.FqNames.APPLICATION
 import com.deliveryhero.whetstone.compiler.GeneratedFileInfo
 import com.deliveryhero.whetstone.compiler.getValue
 import com.squareup.anvil.annotations.MergeComponent
@@ -59,14 +60,21 @@ internal class AppComponentHandler : CodegenHandler {
             .addSuperinterface(applicationComponentCn.nestedClass("Factory"))
             .addAnnotation(Component.Factory::class)
             .build()
+        val appParam = ParameterSpec.builder("application", APPLICATION.asClassName(module))
+            .build()
         val companionObjectSpec = TypeSpec.companionObjectBuilder("Default")
-            .addSuperinterface(
-                generatedComponentCn.nestedClass("Factory"),
-                CodeBlock.of(
-                    "%T.factory()",
-                    generatedComponentCn.peerClass("Dagger${generatedComponentCn.simpleName}")
-                )
-            )
+            .addSuperinterface(generatedComponentCn.nestedClass("Factory"))
+            .addFunction(
+                FunSpec.builder("create")
+                    .addParameter(appParam)
+                    .returns(generatedComponentCn)
+                    .addModifiers(KModifier.PUBLIC, KModifier.OVERRIDE)
+                    .addCode(CodeBlock.of(
+                        "return DaggerGeneratedApplicationComponent.factory().create(%L) as %T",
+                        appParam.name,
+                        generatedComponentCn,
+                    ))
+                    .build())
             .build()
 
         val appComponentSpec = TypeSpec.interfaceBuilder(generatedComponentCn)

--- a/whetstone/src/main/java/com/deliveryhero/whetstone/app/ContributesAppInjector.kt
+++ b/whetstone/src/main/java/com/deliveryhero/whetstone/app/ContributesAppInjector.kt
@@ -36,8 +36,11 @@ import com.deliveryhero.whetstone.meta.AutoInjectorBinding
  *    @Component.Factory
  *    public interface Factory : ApplicationComponent.Factory
  *
- *    public companion object Default : Factory by DaggerGeneratedApplicationComponent.factory()
- * }
+ *    public companion object Default : Factory {
+ *      public override fun create(application: Application): GeneratedApplicationComponent =
+ *         DaggerGeneratedApplicationComponent.factory().create(application) as
+ *         GeneratedApplicationComponent
+ *    }
  * ```
  *
  * This can be very handy for quickly bootstrapping the DI setup. It is ideal for cases where the

